File: Seori/Programmers/사라지는 발판.py
def solution(board, aloc, bloc):

    direction = [(0, -1), (-1, 0), (1, 0), (0, 1)]
    
    def dfs(board, my_loc, your_loc, turn):
        # [1] 재귀 종료 조건. 지금 내 발판이 0인 경우 패배 확정이다.
        x, y = my_loc
        if board[x][y] == 0:
            return turn, 0
        
        # [2] 4방향을 탐색하며 가능한 경우 재귀를 시작한다.
        winner = False
        max_move = 0
        min_move = float('inf')
        for dx, dy in direction:
            nx, ny = x + dx, y + dy
            if 0 <= nx < len(board) and 0 <= ny < len(board[0]) and board[nx][ny] == 1:
                
                # [3] 움직이기 위해 지금 발판을 0처리 해주고,
                board[x][y] = 0
                loser, move = dfs(board, your_loc, [nx, ny], not turn) # 재귀를 할 땐 플레이어 순서를 바꿔준다.
                # [4] 재귀 탐색이 끝나면 발판을 다시 1 처리해준다.(백트래킹)
                board[x][y] = 1
                
                # [5] 지금이 진 사람의 턴이라면 최대한 오래 버티도록 플레이
                if turn == loser:
                    max_move = max(max_move, move + 1)
                # [6] 지금이 이긴 사람의 턴이라면 최대한 빨리 이기도록 플레이
                else:
                    winner = True
                    min_move = min(min_move, move + 1)
        
        
        # [7] 승/패에 맞게 움직임 횟수를 반환
        if winner:
            return (not turn, min_move)
        else:
            return (turn, max_move)
    
    return dfs(board, aloc, bloc, False)[1]

# Test case 1
board = [[1, 1, 1], [1, 1, 1], [1, 1, 1]]
aloc = [1, 0]
bloc = [1, 2]
print(solution(board, aloc, bloc))  # Expected output: 5

File: Gombang/Programmers/사라지는발판.cpp
#include <string>
#include <vector>

using namespace std;

// 좌표를 나타내는 구조체
struct Point {
    int x, y;
};
vector<vector<int>> map; // 맵 전역변수화
int n, m; // n : 세로, m : 가로

// 상하좌우
int dx[] = { -1,1,0,0 };
int dy[] = { 0,0,-1,1 };

// me : 현재 턴인 사람 // you : 상대방
int dfs(Point me, Point you) {
    int x = me.x;
    int y = me.y;
    if (map[x][y] == 0) return 0; // 발판이 사라졌다면 0 반환

    int result = 0; // 최종적인 턴 수
    for (int i = 0; i < 4; i++) {
        int nx = x + dx[i];
        int ny = y + dy[i];
        if (nx < 0 || ny < 0 || nx >= n || ny >= m || map[nx][ny] == 0) continue;

        map[x][y] = 0; // 전에 서있던 곳을 이동 불가능하게 만듦

        // 여기서 상대방 턴이기 때문에 매개변수로 dfs(you,me) 순서로 들어간다.
        int val = dfs(you, { nx,ny }) + 1; // 턴 수 + 1
        map[x][y] = 1; // 사용한것을 원상 복구

        // 지금까지 모두 진 경우고, 이번에 이겼을 때
        if (val % 2 == 1 && result % 2 == 0) result = val; // 바로 이긴걸로 바꿔줌

        // 지금까지도 졌고, 이 경우도 진 경우 -> 최대한 많이 움직인다.
        else if (val % 2 == 0 && result % 2 == 0) result = max(result, val);

        // 지금까지도 이겼고, 이 경우도 이긴 경우 -> 최대한 적게 움직인다.
        else if (val % 2 == 1 && result % 2 == 1) result = min(result, val);
    }

    return result;
}
int solution(vector<vector<int>> board, vector<int> aloc, vector<int> bloc) {
    int answer = -1;

    map = board;
    n = board.size();
    m = board[0].size();

    Point a = { aloc[0],aloc[1] };
    Point b = { bloc[0],bloc[1] };

    answer = dfs(a, b);
    return answer;
}